rhp/v4/server.go

```diff
@@ -0,0 +1,1110 @@
+		return fmt.Errorf("failed to lock contract: %w", err)
+	}
+	unlock()
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCLatestRevisionResponse{
+		Contract: state.Revision,
+	})
+}
+
+func (s *Server) handleRPCSectorRoots(stream net.Conn) error {
+	var req rhp4.RPCSectorRootsRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	state, unlock, err := s.lockContractForRevision(req.ContractID)
+	if err != nil {
+		return fmt.Errorf("failed to lock contract: %w", err)
+	}
+	defer unlock()
+
+	// validate the request fields
+	settings := s.settings.RHP4Settings()
+	if err := req.Validate(s.hostKey.PublicKey(), state.Revision, settings.MaxSectorBatchSize); err != nil {
+		return rhp4.NewRPCError(rhp4.ErrorCodeBadRequest, err.Error())
+	}
+	prices := req.Prices
+
+	// update the revision
+	revision, usage, err := rhp4.ReviseForSectorRoots(state.Revision, prices, req.Length)
+	if err != nil {
+		return fmt.Errorf("failed to revise contract: %w", err)
+	}
+
+	// validate the renter's signature
+	cs := s.chain.TipState()
+	sigHash := cs.ContractSigHash(revision)
+	if !state.Revision.RenterPublicKey.VerifyHash(sigHash, req.RenterSignature) {
+		return rhp4.ErrInvalidSignature
+	}
+
+	// sign the revision
+	revision.HostSignature = s.hostKey.SignHash(sigHash)
+
+	// update the contract
+	err = s.contractor.ReviseV2Contract(req.ContractID, revision, state.Roots, usage)
+	if err != nil {
+		return fmt.Errorf("failed to revise contract: %w", err)
+	}
+
+	roots := state.Roots[req.Offset : req.Offset+req.Length]
+	proof := rhp4.BuildSectorRootsProof(state.Roots, req.Offset, req.Offset+req.Length)
+
+	// send the response
+	return rhp4.WriteResponse(stream, &rhp4.RPCSectorRootsResponse{
+		Proof:         proof,
+		Roots:         roots,
+		HostSignature: revision.HostSignature,
+	})
+}
+
+func (s *Server) handleRPCAccountBalance(stream net.Conn) error {
+	var req rhp4.RPCAccountBalanceRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	balance, err := s.contractor.AccountBalance(req.Account)
+	if err != nil {
+		return fmt.Errorf("failed to get account balance: %w", err)
+	}
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCAccountBalanceResponse{
+		Balance: balance,
+	})
+}
+
+func (s *Server) handleRPCFormContract(stream net.Conn) error {
+	var req rhp4.RPCFormContractRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	ourKey := s.hostKey.PublicKey()
+	settings := s.settings.RHP4Settings()
+	tip := s.chain.Tip()
+	if err := req.Validate(ourKey, tip, settings.MaxCollateral, settings.MaxContractDuration); err != nil {
+		return err
+	}
+	prices := req.Prices
+
+	fc, usage := rhp4.NewContract(prices, req.Contract, ourKey, settings.WalletAddress)
+	formationTxn := types.V2Transaction{
+		MinerFee:      req.MinerFee,
```

Comment:
I thought we talked about eliminating the fee from the request and having the host fund the fee in exchange for the renter funding the `ContractPrice`?
It doesn't matter too much but it feels like the host would have the better estimate of network fees considering that it is running 24/7

Reply:
Since the host only validates the fee is above 0 and the transaction is valid, it didn't seem necessary to change the fee payer too. The host will accept the contract and retry the transaction until it's confirmed or the reject height hits. If there's a period of high fees, the formation transaction will probably be accepted later.